dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantMemoryStore.cs

```diff
@@ -0,0 +1,369 @@
+    }
+
+    /// <inheritdoc/>
+    public async Task DeleteCollectionAsync(string collectionName, CancellationToken cancel = default)
+    {
+        if (!await this._qdrantClient.DoesCollectionExistAsync(collectionName, cancel: cancel))
+        {
+            await this._qdrantClient.DeleteCollectionAsync(collectionName, cancel: cancel);
+        }
+    }
+
+    public async Task<string> UpsertAsync(string collectionName, MemoryRecord record, CancellationToken cancel = default)
+    {
+        var vectorData = await this.ConvertFromMemoryRecordAsync(collectionName, record, cancel);
+
+        if (vectorData == null)
+        {
+            throw new VectorDbException(VectorDbException.ErrorCodes.FailedToConvertMemoryRecordToQdrantVectorRecord, $"Failed to convert MemoryRecord to QdrantVectorRecord");
+        }
+
+        try
+        {
+            await this._qdrantClient.UpsertVectorsAsync(
+                collectionName,
+                new[] { vectorData },
+                cancel: cancel);
+        }
+        catch (HttpRequestException e)
+        {
+            throw new VectorDbException(VectorDbException.ErrorCodes.FailedToUpsertVectors, $"Failed to upsert, check if the target collection exists.");
+        }
+
+        return vectorData.PointId;
+    }
+
+    public async IAsyncEnumerable<string> UpsertBatchAsync(string collectionName, IEnumerable<MemoryRecord> record, [EnumeratorCancellation] CancellationToken cancel = default)
+    {
+        var tasks = Task.WhenAll(record.Select(async r => await this.ConvertFromMemoryRecordAsync(collectionName, r, cancel)));
+        var vectorData = await tasks;
+
+        try
+        {
+            await this._qdrantClient.UpsertVectorsAsync(
+                collectionName,
+                vectorData,
+                cancel: cancel);
+        }
+        catch (HttpRequestException e)
+        {
+            throw new VectorDbException(VectorDbException.ErrorCodes.FailedToUpsertVectors, $"Failed to upsert, check if the target collection exists.");
+        }
+
+        foreach (var v in vectorData)
+        {
+            yield return v.PointId;
+        }
+    }
+
+    public async Task<MemoryRecord?> GetAsync(string collectionName, string key, CancellationToken cancel = default)
+    {
+        try
+        {
+            var vectorData = await this._qdrantClient.GetVectorByPayloadIdAsync(collectionName, key, cancel: cancel);
+            if (vectorData != null)
+            {
+                return MemoryRecord.FromJson(
+                    json: vectorData.GetSerializedPayload(),
+                    embedding: new Embedding<float>(vectorData.Embedding),
+                    key: vectorData.PointId);
+            }
+            else
+            {
+                return null;
+            }
+        }
+        catch (Exception ex)
+        {
+            throw new VectorDbException($"Failed to get vector data from Qdrant {ex.Message}");
```

Comment:
Please include the original exception as an innerexception.

Reply:
Same comment for all methods in this file.

Reply:
This is the current pattern for the kernel. VectorDBException.cs is structured the same way as KernelException.cs. Knowing that, do you mind keeping things the way they are?

Reply:
Per our conversation, let's include innerexceptions in these throws.

Reply:
done